programs/drift/src/state/lp_pool.rs

```diff
@@ -0,0 +1,1462 @@
+                    ),
+                ),
+            )
+        }
+    }
+
+    pub fn record_mint_redeem_fees(&mut self, amount: i64) -> DriftResult {
+        self.total_mint_redeem_fees_paid = self
+            .total_mint_redeem_fees_paid
+            .safe_add(amount.cast::<i128>()?)?;
+        Ok(())
+    }
+}
+
+#[zero_copy(unsafe)]
+#[derive(Default, Eq, PartialEq, Debug, BorshDeserialize, BorshSerialize)]
+#[repr(C)]
+pub struct BLPosition {
+    /// The scaled balance of the position. To get the token amount, multiply by the cumulative deposit/borrow
+    /// interest of corresponding market.
+    /// precision: token precision
+    pub scaled_balance: u128,
+    /// The cumulative deposits/borrows a user has made into a market
+    /// precision: token mint precision
+    pub cumulative_deposits: i64,
+    /// The market index of the corresponding spot market
+    pub market_index: u16,
+    /// Whether the position is deposit or borrow
+    pub balance_type: SpotBalanceType,
+    pub padding: [u8; 5],
+}
+
+impl SpotBalance for BLPosition {
+    fn market_index(&self) -> u16 {
+        self.market_index
+    }
+
+    fn balance_type(&self) -> &SpotBalanceType {
+        &self.balance_type
+    }
+
+    fn balance(&self) -> u128 {
+        self.scaled_balance as u128
+    }
+
+    fn increase_balance(&mut self, delta: u128) -> DriftResult {
+        self.scaled_balance = self.scaled_balance.safe_add(delta)?;
+        Ok(())
+    }
+
+    fn decrease_balance(&mut self, delta: u128) -> DriftResult {
+        self.scaled_balance = self.scaled_balance.safe_sub(delta)?;
+        Ok(())
+    }
+
+    fn update_balance_type(&mut self, balance_type: SpotBalanceType) -> DriftResult {
+        self.balance_type = balance_type;
+        Ok(())
+    }
+}
+
+impl BLPosition {
+    pub fn get_token_amount(&self, spot_market: &SpotMarket) -> DriftResult<u128> {
+        get_token_amount(self.scaled_balance, spot_market, &self.balance_type)
+    }
+}
+
+#[account(zero_copy(unsafe))]
+#[derive(Default, Debug, BorshDeserialize, BorshSerialize)]
+#[repr(C)]
+pub struct Constituent {
+    /// address of the constituent
+    pub pubkey: Pubkey,
+    pub mint: Pubkey,
+    pub lp_pool: Pubkey,
+    pub token_vault: Pubkey,
+
+    /// total fees received by the constituent. Positive = fees received, Negative = fees paid
+    pub total_swap_fees: i128,
+
+    /// spot borrow-lend balance for constituent
+    pub spot_balance: BLPosition, // should be in constituent base asset
+
+    /// max deviation from target_weight allowed for the constituent
+    /// precision: PERCENTAGE_PRECISION
+    pub max_weight_deviation: i64,
+    /// min fee charged on swaps to/from this constituent
+    /// precision: PERCENTAGE_PRECISION
+    pub swap_fee_min: i64,
+    /// max fee charged on swaps to/from this constituent
+    /// precision: PERCENTAGE_PRECISION
+    pub swap_fee_max: i64,
+
+    /// ata token balance in token precision
+    pub token_balance: u64,
```

Comment:
we probably want an invariant that makes sure this always matches whats in token account (good way to catch bugs)